File: src/riscv_sim/memory.py
"""A simple byte-addressable, little-endian memory model."""

from .errors import MemoryAccessError


class Memory:
    """Fixed-size byte-addressable memory with little-endian word access."""

    def __init__(self, size: int = 1 << 20) -> None:
        self.size = size
        self._data = bytearray(size)

    def _check_range(self, address: int, length: int) -> None:
        if address < 0 or address + length > self.size:
            raise MemoryAccessError(
                f"address 0x{address:08x} (length {length}) is out of bounds "
                f"for memory of size {self.size} bytes"
            )

    def read_bytes(self, address: int, length: int) -> bytes:
        self._check_range(address, length)
        return bytes(self._data[address:address + length])

    def write_bytes(self, address: int, data: bytes) -> None:
        self._check_range(address, len(data))
        self._data[address:address + len(data)] = data

    def read_byte(self, address: int, signed: bool = False) -> int:
        value = self.read_bytes(address, 1)[0]
        return value - 256 if signed and value & 0x80 else value

    def read_half(self, address: int, signed: bool = False) -> int:
        value = int.from_bytes(self.read_bytes(address, 2), "little")
        return value - 65536 if signed and value & 0x8000 else value

    def read_word(self, address: int) -> int:
        return int.from_bytes(self.read_bytes(address, 4), "little")

    def write_byte(self, address: int, value: int) -> None:
        self.write_bytes(address, (value & 0xFF).to_bytes(1, "little"))

    def write_half(self, address: int, value: int) -> None:
        self.write_bytes(address, (value & 0xFFFF).to_bytes(2, "little"))

    def write_word(self, address: int, value: int) -> None:
        self.write_bytes(address, (value & 0xFFFFFFFF).to_bytes(4, "little"))

    def load_bytes(self, address: int, data: bytes) -> None:
        """Load a block of raw bytes (e.g. a program image) into memory."""
        self.write_bytes(address, data)

File: src/memory.cpp
#include "riscv_sim/memory.hpp"

#include <algorithm>
#include <cstdio>

#include "riscv_sim/errors.hpp"

namespace riscv_sim {

void Memory::check_range(std::uint32_t address, std::size_t length) const {
    if (static_cast<std::size_t>(address) + length > size_) {
        char buf[128];
        std::snprintf(buf, sizeof(buf),
                       "address 0x%08x (length %zu) is out of bounds for memory of size %zu bytes",
                       address, length, size_);
        throw MemoryAccessError(buf);
    }
}

std::vector<std::uint8_t> Memory::read_bytes(std::uint32_t address, std::size_t length) const {
    check_range(address, length);
    return std::vector<std::uint8_t>(data_.begin() + address, data_.begin() + address + length);
}

void Memory::write_bytes(std::uint32_t address, const std::uint8_t* data, std::size_t length) {
    check_range(address, length);
    std::copy(data, data + length, data_.begin() + address);
}

std::int32_t Memory::read_byte(std::uint32_t address, bool signed_) const {
    auto bytes = read_bytes(address, 1);
    std::uint8_t value = bytes[0];
    if (signed_ && (value & 0x80)) {
        return static_cast<std::int32_t>(value) - 256;
    }
    return value;
}

std::int32_t Memory::read_half(std::uint32_t address, bool signed_) const {
    auto bytes = read_bytes(address, 2);
    std::uint32_t value = static_cast<std::uint32_t>(bytes[0]) | (static_cast<std::uint32_t>(bytes[1]) << 8);
    if (signed_ && (value & 0x8000)) {
        return static_cast<std::int32_t>(value) - 65536;
    }
    return static_cast<std::int32_t>(value);
}

std::uint32_t Memory::read_word(std::uint32_t address) const {
    auto bytes = read_bytes(address, 4);
    return static_cast<std::uint32_t>(bytes[0]) | (static_cast<std::uint32_t>(bytes[1]) << 8) |
           (static_cast<std::uint32_t>(bytes[2]) << 16) | (static_cast<std::uint32_t>(bytes[3]) << 24);
}

void Memory::write_byte(std::uint32_t address, std::uint32_t value) {
    std::uint8_t byte = static_cast<std::uint8_t>(value & 0xFF);
    write_bytes(address, &byte, 1);
}

void Memory::write_half(std::uint32_t address, std::uint32_t value) {
    std::uint16_t half = static_cast<std::uint16_t>(value & 0xFFFF);
    std::uint8_t bytes[2] = {static_cast<std::uint8_t>(half & 0xFF), static_cast<std::uint8_t>(half >> 8)};
    write_bytes(address, bytes, 2);
}

void Memory::write_word(std::uint32_t address, std::uint32_t value) {
    std::uint8_t bytes[4] = {
        static_cast<std::uint8_t>(value & 0xFF),
        static_cast<std::uint8_t>((value >> 8) & 0xFF),
        static_cast<std::uint8_t>((value >> 16) & 0xFF),
        static_cast<std::uint8_t>((value >> 24) & 0xFF),
    };
    write_bytes(address, bytes, 4);
}

void Memory::load_bytes(std::uint32_t address, const std::vector<std::uint8_t>& data) {
    write_bytes(address, data.data(), data.size());
}

}  // namespace riscv_sim
